### PR TITLE
Make timeout logic transport agnostic

### PR DESCRIPTION
**Pull request by halter73**
#1586 

@davidfowl @pakrym

### CODE REVIEW

**Comment by halter73** [comment]
The perf impact in end-to-end benchmarks seems negligible. It's hard to do before-and-after micro benchmarks since it's difficult to isolate the old libuv-based timeout logic. Here are some functional results for keep-alive and non keep-alive connections:

### Keep-Alive Results (Dev)
```
shalter@f4-linux:~$ wrk -H 'Host: 10.1.2.6' -H 'Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7' -H 'Connection: keep-alive' --latency -d 15 -c 256 --timeout 8 -t 32 http://10.1.2.6:5000/plaintext -s ~/pipeline.lua -- 16
Running 15s test @ http://10.1.2.6:5000/plaintext
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     6.02ms   13.02ms 238.98ms   98.55%
    Req/Sec    17.28k     3.37k   26.56k    90.31%
  Latency Distribution
     50%    4.14ms
     75%    6.87ms
     90%   10.23ms
     99%   25.97ms
  8189595 requests in 14.73s, 1.01GB read
  Socket errors: connect 0, read 0, write 0, timeout 765
Requests/sec: 556151.57
Transfer/sec:     70.01MB
shalter@f4-linux:~$ wrk -H 'Host: 10.1.2.6' -H 'Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7' -H 'Connection: keep-alive' --latency -d 15 -c 256 --timeout 8 -t 32 http://10.1.2.6:5000/plaintext -s ~/pipeline.lua -- 16
Running 15s test @ http://10.1.2.6:5000/plaintext
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     5.08ms    3.79ms  44.28ms   73.95%
    Req/Sec    16.81k     2.97k   24.48k    88.86%
  Latency Distribution
     50%    4.21ms
     75%    6.88ms
     90%   10.14ms
     99%   17.30ms
  8251266 requests in 14.77s, 1.01GB read
  Socket errors: connect 0, read 0, write 0, timeout 768
Requests/sec: 558626.17
Transfer/sec:     70.32MB
shalter@f4-linux:~$ wrk -H 'Host: 10.1.2.6' -H 'Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7' -H 'Connection: keep-alive' --latency -d 15 -c 256 --timeout 8 -t 32 http://10.1.2.6:5000/plaintext -s ~/pipeline.lua -- 16
Running 15s test @ http://10.1.2.6:5000/plaintext
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     5.13ms    3.75ms  46.36ms   72.28%
    Req/Sec    16.69k     3.01k   24.40k    88.26%
  Latency Distribution
     50%    4.25ms
     75%    7.00ms
     90%   10.33ms
     99%   16.87ms
  8159121 requests in 14.71s, 1.00GB read
  Socket errors: connect 0, read 0, write 0, timeout 768
Requests/sec: 554829.71
Transfer/sec:     69.84MB
```
### Keep-Alive Results (This PR)
```
shalter@f4-linux:~$ wrk -H 'Host: 10.1.2.6' -H 'Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7' -H 'Connection: keep-alive' --latency -d 15 -c 256 --timeout 8 -t 32 http://10.1.2.6:5000/plaintext -s ~/pipeline.lua -- 16
Running 15s test @ http://10.1.2.6:5000/plaintext
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     6.11ms   12.88ms 234.95ms   98.31%
    Req/Sec    17.11k     3.35k   28.64k    90.24%
  Latency Distribution
     50%    4.19ms
     75%    6.93ms
     90%   10.52ms
     99%   27.16ms
  8118293 requests in 14.72s, 1.00GB read
  Socket errors: connect 0, read 0, write 0, timeout 768
Requests/sec: 551484.42
Transfer/sec:     69.42MB
shalter@f4-linux:~$ wrk -H 'Host: 10.1.2.6' -H 'Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7' -H 'Connection: keep-alive' --latency -d 15 -c 256 --timeout 8 -t 32 http://10.1.2.6:5000/plaintext -s ~/pipeline.lua -- 16
Running 15s test @ http://10.1.2.6:5000/plaintext
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     4.99ms    3.85ms  65.35ms   76.53%
    Req/Sec    17.26k     3.00k   22.96k    90.18%
  Latency Distribution
     50%    4.13ms
     75%    6.72ms
     90%   10.12ms
     99%   19.68ms
  8462770 requests in 14.75s, 1.04GB read
  Socket errors: connect 0, read 0, write 0, timeout 765
Requests/sec: 573830.41
Transfer/sec:     72.24MB
shalter@f4-linux:~$ wrk -H 'Host: 10.1.2.6' -H 'Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7' -H 'Connection: keep-alive' --latency -d 15 -c 256 --timeout 8 -t 32 http://10.1.2.6:5000/plaintext -s ~/pipeline.lua -- 16
Running 15s test @ http://10.1.2.6:5000/plaintext
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     5.05ms    3.82ms  53.87ms   74.78%
    Req/Sec    17.14k     2.98k   22.72k    88.47%
  Latency Distribution
     50%    4.14ms
     75%    6.79ms
     90%   10.19ms
     99%   17.54ms
  8408359 requests in 14.76s, 1.03GB read
  Socket errors: connect 0, read 0, write 0, timeout 767
Requests/sec: 569773.70
Transfer/sec:     71.73MB
```

### 1 Request Per Connection Result (Dev)
```
shalter@f4-linux:~$ wrk -H 'Host: 10.1.2.6:5000' -H 'Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7' -H 'Connection: close' --latency -d 2 -c 256 --timeout 8 -t 32 http://10.1.2.6:5000/plaintext
Running 2s test @ http://10.1.2.6:5000/plaintext
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    15.11ms   35.32ms 225.57ms   93.47%
    Req/Sec   628.71    410.66     3.53k    83.16%
  Latency Distribution
     50%    5.23ms
     75%    9.40ms
     90%   18.31ms
     99%  202.63ms
  30878 requests in 1.98s, 4.45MB read
  Socket errors: connect 0, read 0, write 0, timeout 67
Requests/sec:  15607.29
Transfer/sec:      2.25MB
```

### 1 Request Per Connection Result (This PR)
```
shalter@f4-linux:~$ wrk -H 'Host: 10.1.2.6:5000' -H 'Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7' -H 'Connection: close' --latency -d 2 -c 256 --timeout 8 -t 32 http://10.1.2.6:5000/plaintext
Running 2s test @ http://10.1.2.6:5000/plaintext
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    16.25ms   37.15ms 245.27ms   93.21%
    Req/Sec   630.07    279.73     1.37k    69.29%
  Latency Distribution
     50%    4.34ms
     75%   12.09ms
     90%   24.50ms
     99%  204.91ms
  32362 requests in 2.09s, 4.66MB read
Requests/sec:  15476.32
Transfer/sec:      2.23MB
```

The 1 Request Per Connection are short due to the client running out of ephemeral ports.

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/DateHeaderValueManager.cs [hunk 1]
```diff
@@ -12,20 +12,17 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
     /// <summary>
     /// Manages the generation of the date header value.
     /// </summary>
-    public class DateHeaderValueManager : IDisposable
+    public class DateHeaderValueManager : ITick, IDisposable
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/DateHeaderValueManager.cs]: `ITickHandler`?

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/DateHeaderValueManager.cs [hunk 2]
```diff
@@ -107,62 +89,32 @@ public void Dispose()
...
+        // Called by the Timer (background) thread
+        public void Tick(DateTimeOffset now)
         {
-            if (!_isDisposed)
+            if (_timerIsRunning)
             {
-                lock (_timerLocker)
+                SetDateValues(now);
+
+                if (_hadRequestsSinceLastTimerTick)
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/DateHeaderValueManager.cs]: Does `_hadRequestsSinceLastTimerTick` make sense here? We are only preventing header value from encoding when no requests are hitting server, hardly worth additional logic.

File: test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/PipeOptionsTests.cs [hunk 3]
```diff
@@ -7,6 +7,7 @@
 using Microsoft.AspNetCore.Testing;
 using Moq;
 using Xunit;
+using FrameConnection = Microsoft.AspNetCore.Server.Kestrel.Core.Internal.FrameConnection;
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/PipeOptionsTests.cs]: ?

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Infrastructure/Heartbeat.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        private readonly Timer _timer;
+        private int _executingOnBeat;
+
+        public Heartbeat(IEnumerable<ITick> callbacks, ISystemClock systemClock, IKestrelTrace trace)
+            : this(callbacks, systemClock, Interval, trace)
+        {
+        }
+
+        // For testing
+        internal Heartbeat(IEnumerable<ITick> callbacks, ISystemClock systemClock, TimeSpan interval, IKestrelTrace trace)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Infrastructure/Heartbeat.cs]: Make it an array so we don't allocate every time we run tick

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Infrastructure/Heartbeat.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure
+{
+    public class Heartbeat : IDisposable
+    {
+        public static readonly TimeSpan Interval = TimeSpan.FromMilliseconds(1000);
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Infrastructure/Heartbeat.cs]: `FromSeconds(1)`?

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Infrastructure/Heartbeat.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            _callbacks = callbacks;
+            _interval = interval;
+            _systemClock = systemClock;
+            _trace = trace;
+            _timer = new Timer(OnBeat, state: this, dueTime: TimeSpan.Zero, period: _interval);
+        }
+        
+        // Called by the Timer (background) thread
+        private void OnBeat(object state)
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Infrastructure/Heartbeat.cs]: Add a test which calls `OnBeat` twice blocking in the `Tick` so you can test non reentrancy and `TimerSlow` log message.

**Comment by pakrym** [comment]
LGTM

**Pull request merged**